fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordCursor.java

```diff
@@ -952,4 +953,25 @@ public boolean isLimitReached() {
         final AtomicReference<U> holder = new AtomicReference<>(identity);
         return forEachResult(result -> holder.set(accumulator.apply(holder.get(), result.get()))).thenApply(vignore -> holder.get());
     }
-}
+
+    /**
+     * A reduce variant which allows stopping the processing when the value
+     * has been successfully reduced
+     * @param identity initial value for reduction
+     * @param accumulator function that takes previous reduces value and computes new value by combining with each record
+     * @param stopCondition predicate which runs against the accumulated result before fetching new records
+     * @param <U> the result type of the reduction
+     * @return a future that completes to the result of reduction
+     */
+    @Nullable
+    default <U> CompletableFuture<U> reduce(U identity, BiFunction<U, ? super T, U> accumulator, Predicate<U> stopCondition) {
+        final AtomicReference<U> holder = new AtomicReference<>(identity);
+        return AsyncUtil.whileTrue(() -> onNext().thenApply(result -> {
+            U nextResult = accumulator.apply(holder.get(), result.get());
```

Comment:
This will get an error (`Tried to call get() on a RecordCoreResult that did not have a next value.`) if the result doesn't have a next.

Reply:
Indeed, this was a rough translation of our in-house Clojure version and got things mixed up- fixed in latest commit